hphp/runtime/vm/jit/vasm-flags.cpp

```diff
@@ -0,0 +1,215 @@
+#undef O
+  }
+  not_reached();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/*
+ * For each use of an SF, return the StatusFlags bits needed by the use's CC.
+ *
+ * Specializations are below, which can additionally change the CC if the
+ * platform doesn't have a simple CC mapping for the given def and use.
+ */
+template<typename Inst>
+Vflags annotateSFUses(Env& env, Inst& inst) {
+  return static_cast<Vflags>(required_flags(inst.cc));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+Vflags annotateSFUses(Env& env, jcc& inst) {
+  if (env.sf_def_uses[inst.sf].def->op == Vinstr::ucomisd) {
+    switch (inst.cc) {
+    case CC_AE: // CC_NB
+      inst.cc = CC_GE;
+      break;
+    default:
+      break;
+    }
+  }
+  return static_cast<Vflags>(required_flags(inst.cc));
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+/*
+ * Map this uses CC in any platform-specific way and return the platform-
+ * specific status flags bits that are used for the CC.
+ */
+Vflags annotateSFUses(Env& env, Vinstr* inst) {
+  switch (inst->op) {
+#define O(name, imms, ...) \
+    case Vinstr::name: { \
+      auto& i = inst->name##_; (void)i; \
+      imms \
+      break; \
+  }
+#define Inone
+#define IF(fl)
+#define IC(cc) return annotateSFUses(env, i);
+#define I(x)
+    VASM_OPCODES
+#undef I
+#undef IC
+#undef IF
+#undef Inone
+#undef O
+  }
+  not_reached();
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+}
+
+/*
+ * Map SF pass for a Vunit.
+ */
+void annotateSFUses(Vunit& unit) {
+
+  if (!arch_any(Arch::ARM)) return;
+
+  assertx(check(unit));
+  auto& blocks = unit.blocks;
+
+  Env env { unit };
+
+  auto const labels = sortBlocks(unit);
+
+  // Set up Env, only visiting reachable blocks.
+  // Tracks each VregSF to the instruction which def'd it.
+  // Tracks each VregSF to a list of each instruction which uses it.
+  for (auto const b : labels) {
+    assertx(!blocks[b].code.empty());
+    for (auto& inst : blocks[b].code) {
+      visitDefs(unit, inst, [&] (Vreg r, Width w) {
+        if (w == Width::Flags) {
+          env.sf_def_uses[r].def = &inst;
+        }
+      });
+      visitUses(unit, inst, [&] (Vreg r, Width w) {
+        if (w == Width::Flags) {
+          env.sf_def_uses[r].uses.push_back(&inst);
+        }
+      });
+    }
+  };
+
+  // Calls annotateSFUses on each
+  for (auto&& kv : env.sf_def_uses) {
+    DEBUG_ONLY auto r = kv.first;
+    auto du = kv.second;
+
+    // For each instruction that defs an sf, initially clear the flag bits.
+    Vflags flags = 0x00;
+
+    // Call annotateSFUses for each use. This will a) update the instruction's
+    // condition code (cc) based on any platform specific mapping b) update
+    // the flag bits (c) for this VregSF def based on the instruction's cc.
+    for (auto const u : du.uses) {
+      flags |= annotateSFUses(env, u);
```

Comment:
You've defined `Vflags` as a sort of generic type that different platforms can use as they wish—but there is an implicit requirement here that it be coherently `|`-able as an integer.  I don't think it's worth generalizing until there are more users, but at least you should document this behavior when defining `Vflags`.